### PR TITLE
Deprecated proc_macro doesn't trigger warning on build library

### LINKED ISSUE #65189: deprecated proc_macro trigger warning on build library
Use `deprecated` attribute on a procedural macro function doesn't work like expected: it triggers a warning message also when library is building and not just in call place.

I know that the `deprecate` attribute support is not landed in stable but this warning is present in all channels (stable, beta and nightly).

```
  stable-x86_64-unknown-linux-gnu unchanged - rustc 1.38.0 (625451e37 2019-09-23)
    beta-x86_64-unknown-linux-gnu unchanged - rustc 1.39.0-beta.5 (fa5c2f3e5 2019-10-02)
   nightly-x86_64-unknown-linux-gnu updated - rustc 1.40.0-nightly (421bd77f4 2019-10-06)
```
Compiling lib simple by `cargo +beta build` raise follow warning:

```
   Compiling deprecete_proc_macro v0.1.0 (/home/michele/temp/ppp/deprecete_proc_macro)
warning: use of deprecated item 'deprecable_fn'
 --> src/lib.rs:6:1
  |
6 | pub fn deprecable_fn(input: proc_macro::TokenStream) -> proc_macro::TokenStream { input }
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(deprecated)]` on by default

warning: use of deprecated item 'deprecable_derive'
  --> src/lib.rs:10:1
   |
10 | pub fn deprecable_derive(input: proc_macro::TokenStream) -> proc_macro::TokenStream { input }
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: use of deprecated item 'deprecable_attr'
  --> src/lib.rs:14:1
   |
14 | pub fn deprecable_attr(_args: proc_macro::TokenStream, input: proc_macro::TokenStream) -> proc_macro::TokenStream { input }
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

    Finished dev [unoptimized + debuginfo] target(s) in 0.70s

```

Sadly this issue make `deprecated` attribute not usable at all for procedural macro.

Follow the code to reproduce this issue:

`Cargo.toml`
```toml
[package]
name = "deprecete_proc_macro"
version = "0.1.0"
authors = ["me <myself@gmail.com>"]
edition = "2018"

[lib]
proc_macro=true
```

`src/lib.rs`
```rust
extern crate proc_macro;

#[deprecated]
#[proc_macro]
pub fn deprecable_fn(input: proc_macro::TokenStream) -> proc_macro::TokenStream { input }

#[deprecated]
#[proc_macro_derive(Deprecable)]
pub fn deprecable_derive(input: proc_macro::TokenStream) -> proc_macro::TokenStream { input }

#[deprecated]
#[proc_macro_attribute]
pub fn deprecable_attr(_args: proc_macro::TokenStream, input: proc_macro::TokenStream) -> proc_macro::TokenStream { input }
```

This issue has been assigned to @XiangQingW via [this comment](https://github.com/rust-lang/rust/issues/65189#issuecomment-539278875).

### PR DESCRIPTION
**Pull request by XiangQingW**
Fix #65189

### CODE REVIEW

File: src/libsyntax_ext/proc_macro_harness.rs [hunk 1]
```diff
@@ -416,6 +416,16 @@ fn mk_decls(
...
         i.attrs.push(cx.attribute(attr));
+
+        let deprecated_attr = attr::mk_nested_word_item(
+            Ident::new(sym::deprecated, span)
+        );
+        let allow_deprecated_attr = attr::mk_list_item(
+            Ident::new(sym::allow, span),
+            vec![deprecated_attr]
+        );
+        i.attrs.push(cx.attribute(allow_deprecated_attr));
```
Comment (Centril) [inline src/libsyntax_ext/proc_macro_harness.rs]: Can you please also update the expansion in the doc comment?

**Comment by Centril** [comment]
cc @petrochenkov

File: src/test/ui/proc-macro/proc-macro-deprecated-attr.rs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+// build-pass
```
Comment (petrochenkov) [inline src/test/ui/proc-macro/proc-macro-deprecated-attr.rs]:
```suggestion
// check-pass
```
It's necessary to do codegen/linking to check deprecation.

File: src/test/ui/proc-macro/proc-macro-deprecated-attr.rs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
```
Comment (petrochenkov) [inline src/test/ui/proc-macro/proc-macro-deprecated-attr.rs]: Could you add `#![deny(deprecated)]` here to make sure that the test fails if the deprecation is actually reported.

File: src/test/ui/proc-macro/proc-macro-deprecated-attr.rs [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+extern crate proc_macro;
+use proc_macro::*;
+
+#[proc_macro]
+#[deprecated(since = "1.0.0", note = "test")]
+pub fn test_compile_without_warning_with_deprecated(_: TokenStream) -> TokenStream {
+    "
+    extern crate proc_macro;
+    fn foo() { }
+    ".parse().unwrap()
```
Comment (petrochenkov) [inline src/test/ui/proc-macro/proc-macro-deprecated-attr.rs]: This can be removed, the body is unimportant and can be just `TokenStream::new()`.

**Comment by petrochenkov** [comment]
Some minor comments.

**Comment by XiangQingW** [comment]
@petrochenkov @Centril Tanks for your help. I amend the pr. Would you mind reviewing it again if you have free time?

**Comment by petrochenkov** [comment]
Thanks!
@bors r+

**Comment by Centril** [comment]
Failed in https://github.com/rust-lang/rust/pull/65723#issuecomment-545589029, @bors r-

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/65484#issuecomment-543134361 said to copy the file rather than create it from scratch because the obscure header
```rust
// force-host
// no-prefer-dynamic
```
is important for cross-compilation.

**Comment by XiangQingW** [comment]
> [#65484 (comment)](https://github.com/rust-lang/rust/pull/65484#issuecomment-543134361) said to copy the file rather than create it from scratch because the obscure header
> 
> ```rust
> // force-host
> // no-prefer-dynamic
> ```
> 
> is important for cross-compilation.
> It should be kept.

Could you please review again? It's important, however, I can't found the document about it.

**Pull request merged**